bob-backend/src/pearl/group.rs

```diff
@@ -569,39 +598,72 @@ impl Group {
         });
     }
 
-    pub async fn delete(&self, key: BobKey) -> Result<u64, Error> {
-        let holders = self.holders.read().await;
+    pub async fn delete(
+        &self,
+        key: BobKey,
+        timestamp_config: StartTimestampConfig,
+        force_delete: bool,
+    ) -> Result<u64, Error> {
+        {
+            let holders = self.holders.read().await;
+            if holders.len() != 0 {
+                let copy = holders.iter().cloned().enumerate().collect();
+                drop(holders);
+                return self.delete_in_holders(copy, key, force_delete).await;
+            }
+        }
+        let holder = self
+            .get_actual_holder(get_current_timestamp(), timestamp_config)
+            .await?;
+        self.delete_in_holders(vec![(holder.0, holder.1)], key, force_delete)
+            .await
+    }
+
+    async fn delete_in_holders(
+        &self,
+        holders: Vec<(ChildId, Holder)>,
+        key: BobKey,
+        force_delete: bool,
+    ) -> Result<u64, Error> {
         let mut total_count = 0;
-        for holder in holders.iter() {
-            let delete = Self::delete_common(holder.clone(), key).await;
+        for holder in holders {
+            let delete = Self::delete_common(holder.1.clone(), key, force_delete).await;
```

Comment:
You should find actual holder by timestamp from `BobMeta` and pass `force_delete` only to that holder

Reply:
If `force_delete` == false, then this stage can be skipped